compiler/IREmitter/IREmitter.cc

```diff
@@ -609,9 +609,17 @@ void emitUserBody(CompilerState &base, cfg::CFG &cfg, const IREmitterContext &ir
                     isTerminated = true;
 
                     auto *var = Payload::varGet(cs, i->what.variable, builder, irctx, bb->rubyBlockId);
+                    bool hasBlockAncestor = false;
+                    int rubyBlockId = bb->rubyBlockId;
 
-                    if (irctx.rubyBlockType[bb->rubyBlockId] == FunctionType::Block) {
-                        IREmitterHelpers::emitReturnFromBlock(cs, cfg, builder, irctx, bb->rubyBlockId, var);
+                    while (rubyBlockId != 0) {
+                        hasBlockAncestor = hasBlockAncestor || irctx.rubyBlockType[rubyBlockId] == FunctionType::Block;
```

Comment:
Might be nice to move the deleted `ENFORCE`s up here? (To be honest, though, I'm not sure how critical they really are.)

```suggestion
                        if (irctx.rubyBlockType[rubyBlockId] == FunctionType::Block) {
                            ENFORCE(!functionTypeNeedsPostprocessing(irctx.rubyBlockType[rubyBlockId]));
                            ENFORCE(!functionTypePushesFrame(irctx.rubyBlockType[rubyBlockId]));
                            hasBlockAncestor = true;
                            break;
                        }
```

Reply:
At least the `functionTypeNeedsPostprocessing` one is reasonable to keep around, since that would imply there's some typechecking we need to do.  I think technically we'd want to assert that for the entire ancestor chain, because otherwise we might inadvertently miss some postprocessing by doing the throw.  Does that change sound good to you?

I think we can drop the `functionTypePushesFrame` one.

Reply:
> I think technically we'd want to assert that for the entire ancestor chain, because otherwise we might inadvertently miss some postprocessing by doing the throw. Does that change sound good to you?

Yep, that's a good point! I'd be confused and bewildered if the assert ended up failing anywhere between the return statement and the enclosing `FunctionType::Block`.

> I think we can drop the functionTypePushesFrame one.

Agreed.